### ISSUE TITLE
SLATE Volume - improve error message when attempting to delete volumes that are being used by an application

### ISSUE THREAD

**Issue by jstidd** [slateci/slate-client-server#116]
When deleting a volume that is being used by an application, the request times out and the client receives an HTML error page from nginx. Handle error and create a better error message.

**Comment by Mansalu** [comment]
In this case, did the volume in question actually get removed from the DB and from Kubernetes despite the odd HTTP response?

**Comment by Mansalu** [comment]
Testing locally...the command just hangs forever.

**Comment by Mansalu** [comment]
So, it seems that PVC deletion within Kubernetes just hangs. We can try force deletion but I suspect the behavior there will be bad as well. Perhaps the correct behavior is to, determine if the volume is bound by an application, warn the user the application will be deleted if the volume is deleted, then proceed with instance deletion followed by PVC deletion.

**Comment by Mansalu** [comment]
A significant issue I see with SLATE offering to delete the instance a Volume is bound to is that there may not be a reliable way to determine what SLATE instance bound the volume. We could, perhaps extend SLATE's app install logic to add a label containing the instance ID which the volume code could collect from `kubectl`. 

I think the easy fix would be to warn the user that the volume could not be deleted as it was tied to an instance of some application, and just abort.

**Comment by Mansalu** [comment]
Related Slack discussion from today

```
Mitchell Steinman  7:50 AM
Related to https://github.com/slateci/slate-client-server/issues/116
SLATE Volumes that are bound by an application cannot be deleted as long as they are bound (this is something that seems to just fail within Kubernetes). In this situation, I would like SLATE to be able to offer to delete the instance and then the volume, but it may be non-trivial for SLATE to determine the instance ID consuming a particular volume, since this all happens roughly at the Kubernetes layer. One idea is to have SLATE app install inject a label into the instance which contains SLATE's instance ID. This could be queried from the pod later, but that seems like a reasonable hefty workaround, though having IDs available as labels within K8s could prove useful for other reasons.
A simple solution would be to detect that the volume is bound, warn the user that it is basically there mess to sort out, and abort the operation. That is unhelpful...but doesn't require anything fancy or new code. (edited) 

jstiddjstidd
#116 SLATE Volume - improve error message when attempting to delete volumes that are being used by an application
When deleting a volume that is being used by an application, the request times out and the client receives an HTML error page from nginx. Handle error and create a better error message.
Assignees
jmcragun
Labels
bug, good first issue, server
<https://github.com/slateci/slate-client-server|slateci/slate-client-server>slateci/slate-client-server | Dec 9th | Added by GitHub
7:52
Maybe I am missing something, and there is a better way for SLATE to determine which instance a particular pod is related to. But I don't think so.

cnweaver  7:54 AM
At least detecting that the volume is in use (and also showing this when listing/inspecting volumes if it isn't there already) seems like it would be a big step in the right direction.
7:55
I suspect it's not possible to always make a meaningful determinination of what is using a given volume, since a user could create a vlume through SLATE and then consume it with a random, manually deployed pod?

Mitchell Steinman  8:00 AM
That's a good point I hadn't considered...I suppose in my hypothetical scenario where there is some label SLATE adds like slateInstance: instance_<blah> it would be possible to detect if a Volume was bound by something SLATE did not deploy, but that presents yet another layer of complexity.
Today, I will just get this to the point where volume status (bound/pending) is correctly reported, and the user is warned when attempting to delete a bound volume.

Mitchell Steinman  9:07 AM
I have a local version of SLATE able to report the Volume Status which I thought would be sufficient, but it turns out, that this field has different behavior depending on the StorageClass :man-facepalming:
9:07
admin@Mitchells-MBP-3 build % ./slate volume list
Name      Group Cluster  StorageClass ID                 Status
a-vol     ms    minikube standard     volume_oO4usDTQTP4 Bound
mv        ms    minikube standard     volume_q67l7Q54ai8 Bound
my-volume ms    minikube standard     volume_o2AXqq8DcmY Bound
vol       ms    minikube standard     volume_vnuHVcDzsf0 Bound
9:08
Some SCs will bind the volume upon PVC creation, others wait until a Pod comes along and tries to mount the PVC.
9:08
So this isn't a reliable way to report if a volume is in use :disappointed:

Mitchell Steinman  9:15 AM
The async request to K8s to remove the PVC mounted by a pod just results in a PVC with infinite "terminating" state and a crashed pod.

Lincoln Bryant  9:15 AM
indeed
9:17
It looks like in relatively recent versions (within the last 2yr..?) of Kubernetes, you should be able to kubectl describe pvc and get a Mounted by field
9:17
im just now catching up but

Mitchell Steinman  9:18 AM
Indeed! :exploding_head:
Mounted By:    osg-hosted-ce-hi-647b8bfb6-6p8lc

Lincoln Bryant  9:18 AM
assuming that's the case, I think SLATE should simply inform the user that the PVC is in use and they should deal with the mounters first, as SLATE might not be able to deal with it.
9:18
https://github.com/kubernetes/kubernetes/pull/65837 here's that PR where it was merged in
clandry94clandry94
#65837 output which pods are in use by a pvc when kubectl describe pvc is run
What this PR does / why we need it:  
This PR implements a feature request to display the pods mounted to a PVC when kubectl describe pvc is ran. We need it since the only way to see which pods are bound to a PVC is by checking the volume mounts of each individual pod in the namespace.
Example Output:
$ kubectl get pods,pvc
NAME     READY  STATUS    RESTARTS   AGE
test-pod   1/1       Running       0                2h

Show more
Assignees
mengqiy, juanvallejo
Labels
approved, cncf-cla: yes, lgtm, release-note, sig/cli, sig/storage, size/M
<https://github.com/kubernetes/kubernetes|kubernetes/kubernetes>kubernetes/kubernetes | Jul 4th, 2018 | Added by GitHub
9:18
of course in k8s's infinite helpfulness it doesn't say anything about what version it was merged in
9:18
you'd think with all of their robots they'd go back thru and tag PRs with a release

Mitchell Steinman  9:20 AM
It's also, helpfully, not available in the JSON output  :thinking_face:

Lincoln Bryant  9:20 AM
/facepalm
9:20
at the bottom of that ticket someone asks if it can be exposed to the Golang API
9:20
wonder if they did it in a hacky way

Mitchell Steinman  9:21 AM
Would it be possible to have getMountPods() exposed to the Golang API? Was looking for this exact method to use but just had to copy and paste it over to my own code. Seems odd to not expose it to the wider API, and just restrict it to kubectl describe. So similar to @ps10's request, expose the "Mounted By" to the PersistentVolumeClaimStatus perhaps
Uhg
9:24
And kubectl describe just doesn't support structured output?
9:24
https://stackoverflow.com/questions/37464518/how-to-format-the-output-of-kubectl-describe-to-json

Stack OverflowStack Overflow
How to format the output of kubectl describe to JSON
kubectl get command has this flag -o to format the output. Is there a similar way to format the output of the kubectl describe command? For example: kubectl describe -o="jsonpath={...}" pods my-...

Lincoln Bryant  9:24 AM
:exploding_head:

Mitchell Steinman  9:25 AM
That field is exactly what I need for this, but it looks like it just isn't exposed in a way that can be programmatically accessed? Insanity.

cnweaver  9:26 AM
Yeah, describe is apparently just meant for human consumption, but it's often hard to track down how it gets the data it displays.
9:28
This sort of thing was why I had always vaguely wished to make the slate API server into a direct client of the Kubernetes REST API, but that would be a lot of work and would bring exposure to API changes which are currently mostly abstracted away by kubectl.

Mitchell Steinman  9:30 AM
Any tricks or flags that you've used to get info that is displayed in kubectl describe but not kubectl get -o=json ? (edited) 

1 reply
Today at 9:57 AMView thread

Mitchell Steinman  9:45 AM
Uhg...I guess something like?
kubectl get pods --namespace nspace -o=json | jq -c \
'.items[] | {name: .metadata.name, namespace: .metadata.namespace, claimName:.spec.volumes[] | select( has ("persistentVolumeClaim") ).persistentVolumeClaim.claimName }'
which is awful

Lincoln Bryant  9:46 AM
well
9:46
bravo for fancy use of jq
9:46
but also
9:46
does that even work though? what if the pod is mounting from another namespace?

Mitchell Steinman  9:46 AM
It's buggy
9:47
I don't think a pod can mount from another namespace? The other issue with the jq hack is the situation where the pod doesn't have any volume.

Lincoln Bryant  9:47 AM
ah
9:48
well, we probably want the client to do the JSON parsing anyhow tho right?
9:48
does SLATE normally use jq for anything? I wouldn't necessarily want to add another implicit dependency just for this
9:48
s/client/server

Mitchell Steinman  9:48 AM
SLATE wouldn't use jq

Lincoln Bryant  9:48 AM
got it
9:48
OK

Mitchell Steinman  9:49 AM
I'm trying to make it work on my CLI before I write any code.

Lincoln Bryant  9:49 AM
yeah gotcha
9:49
so it sounds like you have to iterate through every pod in the namespace, check for a volume, if there is one, check to see if it corresponds to the volume you want to delete?

Mitchell Steinman  9:50 AM
Yeah...it's not great...
9:54
It's roughly what the code in describe that you linked is doing though...
9:55
func getMountPods(c coreclient.PodInterface, pvcName string) ([]api.Pod, error) {
	nsPods, err := c.List(metav1.ListOptions{})
    if err != nil {
		return []api.Pod{}, err
	}
	var pods []api.Pod
	for _, pod := range nsPods.Items {
		pvcs := getPvcs(pod.Spec.Volumes)
		for _, pvc := range pvcs {
			if pvc.PersistentVolumeClaim.ClaimName == pvcName {
				pods = append(pods, pod)
			}
		}
	}
	return pods, nil
}
New

Lincoln Bryant  9:55 AM
ya

Mitchell Steinman  9:56 AM
They never added a field to the actual API object, describe just queries all the pods in the namespace and compiles that info for us.
9:56
Which, I guess, explains why the info is never exposed as JSON.

cnweaver  9:58 AM
Well, it's a bit expensive, but it would be possible for slate to do that same.
9:59
I think it probably shouldn't be done without an explicit need/request from the user, though, or it could get very slow in namespaces like ones with Lincoln's many condor workers.

Mitchell Steinman  10:01 AM
Yeah I was thinking on that..
slate volume list for example would need to generate something like O(n*p) requests to K8s. 'p' requests for each pod in the namespace for each of the 'n' volumes tracked by SLATE.

cnweaver  10:03 AM
volume list could exploit the fact that it knows it's scanning the whole namespace, and share the results of the p pod queries across the n volume lookups.
10:04
Even n+p could be pretty expensive, though, when the user just wanted to glance at some other information.

Mitchell Steinman  10:08 AM
Perhaps volume delete/info should be the only commands that try to make this query. Since it is the only place where the information is of critical importance. If I want to know whether or not my application mounted a volume...I can typically just query the application. A user running volume delete or volume info might not know about the application, so we should expose it there. But I think volume list can probably be left alone? It's just meant to be an overview.
For a single volume you need to make P queries to K8s for P pods in that given namespace. (edited) 
```

**Comment by Mansalu** [comment]
I'm building some changes for this on a branch

https://github.com/slateci/slate-client-server/tree/extend-volumes

**Comment by Mansalu** [comment]
Related to today's extensive discussion: I have a fix that addresses this issue in #119
